### PR TITLE
N-body simulation with conceptual layers

### PR DESCRIPTION
**Pull request by samo-novak**
N-body simulation using Effects and Handlers: a layered abstraction
===================================================================

This is a more conceptual (though at the moment quite inefficient) version of the N-body problem. This is a more handler heavy version, using effects and their handlers to implement both state and concurrency.

The main idea is that the system is separated into the Bodies and the Universe:

The Bodies (sun and planets) each have their own state and handle their own evolution, while interacting with the Universe - however they know nothing about the other objects in existence. The Universe acts as the environment that provides mutual interactions for all the bodies.

Each Body is represented by an infinite loop that requests an interaction and advances its own state accordingly. This way, the Bodies execute concurrently - this demonstrates another use of effect handlers.

In a way, we could see the Bodies as processes/threads and the Universe as the Kernel of an operating system.

Communication between the Bodies and the Universe is described in more detail in the following diagram:

nbody_layered_individual

---

This being the N-body problem, it has fully deterministic mechanics. However, this approach could be useful to model other stuff, e.g. active matter (statmech), living organisms, games,...

I can also imagine it being used for an approximation-based speedup, where far away bodies evolve in larger intervals than bodies close to each other (this may be a bad idea though).

I think it also makes for an interesting example of handler-heavy code.

### TODO:

- [x] Figure out how to get a body's kinetic energy without breaching the conceptual boundary too much (this is a tradeoff between idealism and efficiency it seems, but maybe there's a neat solution)
- [x] Interaction
- [x] Creating bodies
- [x] Running the sim
- [x] Δv → Δp (contributes to conceptual split)
- [x] Nice header, description
- [x] Cleanup

---

[1]: Icons (Sun, Jupiter, Saturn) made by [Freepik](https://www.flaticon.com/authors/freepik "Freepik") from [www.flaticon.com](https://www.flaticon.com/ "Flaticon")
[2]: Icon (Uranus) made by [smalllikeart](https://www.flaticon.com/authors/smalllikeart "smalllikeart") from [www.flaticon.com](https://www.flaticon.com/ "Flaticon")
[3]: Icon (Neptune) made by [turkkub](https://www.flaticon.com/authors/turkkub "turkkub") from [www.flaticon.com](https://www.flaticon.com/ "Flaticon")

### CODE REVIEW

File: handlers/nbody_layered.links [hunk 1]
```diff
@@ -0,0 +1,371 @@
+############################################################################################
+# N-body simulation using Effects and Handlers: layered abstraction                        #
+# =================================================================                        #
+#                                                                                          #
+# TODO fill this part (reference <simple.links>)                                           #
```
Comment (dhil) [inline handlers/nbody_layered.links]: Just complete this TODO, then we can merge.

**Review by dhil — APPROVED** [review]
LGTM.

**Comment by samo-novak** [comment]
This is now finished (and with pretty pictures :laughing:).

**Review by dhil — APPROVED** [review]

**Comment by dhil** [comment]
Thanks!

**Pull request merged**